## Updated homework - Arina L.

The aim of this exercise is performing data manipulations in order to derive the requested insights about the company's sales.

## Data cleaning

In [1]:
import pandas as pd
import matplotlib as plt

In [2]:
df = pd.read_csv('ft.csv')
df_anon = pd.read_csv('ft anon.csv')

In [3]:
df

id                          name                             slug  \
0   3418                 Kit 30 filtri                    kit-30-filtri   
1   3401  Infinity azzurro + 30 filtri  infinity-mask-azzurro-30-filtri   
2   3387   Infinity grigio + 30 filtri   infinity-mask-grigio-30-filtri   
3   3351    Infinity verde + 30 filtri    infinity-mask-verde-30-filtri   
4   3397  Infinity arancio + 30 filtri  infinity-mask-arancio-30-filtri   
..   ...                           ...                              ...   
64  1369   T-shirt Fitness Camo Yellow      t-shirt-fitness-camo-yellow   
65  1362     T-shirt Fitness Camo Grey        t-shirt-fitness-camo-grey   
66  1355     T-shirt Fitness Camo Blue        t-shirt-fitness-camo-blue   
67  1348        3/4 Leggings Camo Grey   3-4-leggings-fitness-camo-grey   
68  1341      3/4 Leggings Camo Yellow         3-4-leggings-camo-yellow   

                                            permalink         date_created  \
0   https://archived.forceteksport.it/product/kit-...  2020-04-28T15:49:52   
1   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:36   
2   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:18   
3   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:09   
4   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:08   
..                                                ...                  ...   
64  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:30   
65  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:30   
66  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:29   
67  https://archived.forceteksport.it/product/3-4-...  2019-12-05T08:56:28   
68  https://archived.forceteksport.it/product/3-4-...  2019-12-05T08:56:27   

       date_created_gmt        date_modified    date_modified_gmt      type  \
0   2020-04-28T15:49:52  2021-08-10T14:37:42  2021-08-10T14:37:42  variable   
1   2020-04-28T10:00:36  2021-09-02T06:31:50  2021-09-02T06:31:50  variable   
2   2020-04-28T10:00:18  2021-09-24T08:24:08  2021-09-24T08:24:08  variable   
3   2020-04-28T10:00:09  2021-10-03T18:05:27  2021-10-03T18:05:27  variable   
4   2020-04-28T10:00:08  2021-09-04T05:55:29  2021-09-04T05:55:29  variable   
..                  ...                  ...                  ...       ...   
64  2019-12-05T08:56:30  2021-02-27T18:35:52  2021-02-27T18:35:52  variable   
65  2019-12-05T08:56:30  2021-06-18T15:59:29  2021-06-18T15:59:29  variable   
66  2019-12-05T08:56:29  2021-02-27T18:35:52  2021-02-27T18:35:52  variable   
67  2019-12-05T08:56:28  2020-10-27T14:10:30  2020-10-27T14:10:30  variable   
68  2019-12-05T08:56:27  2021-01-23T14:58:15  2021-01-23T14:58:15  variable   

     status  ...                                         price_html  \
0   publish  ...  <span class="woocommerce-Price-amount amount">...   
1   publish  ...  <span class="woocommerce-Price-amount amount">...   
2   publish  ...  <span class="woocommerce-Price-amount amount">...   
3   publish  ...  <span class="woocommerce-Price-amount amount">...   
4   publish  ...  <span class="woocommerce-Price-amount amount">...   
..      ...  ...                                                ...   
64  publish  ...  <del aria-hidden="true"><span class="woocommer...   
65  publish  ...  <del aria-hidden="true"><span class="woocommer...   
66  publish  ...  <del aria-hidden="true"><span class="woocommer...   
67  publish  ...  <del aria-hidden="true"><span class="woocommer...   
68  publish  ...  <del aria-hidden="true"><span class="woocommer...   

                       related_ids  \
0   [1459, 1555, 1596, 1341, 1540]   
1   [1569, 1422, 1396, 1555, 1376]   
2   [1434, 1355, 1390, 1396, 1422]   
3   [1569, 1459, 1369, 1404, 1396]   
4   [1404, 1569, 1596, 1411, 1341]   
..                             ...   
64  [1589, 1390, 1562, 1355, 1376]   
65  [1589, 1376, 1417, 1533, 1422]   
66  [1369, 1427, 1589, 1

In [4]:
columns_to_delete = []
for column in df.columns:
    if len(df[column].unique()) == 1:
        columns_to_delete.append(column)
        
print(columns_to_delete)

['type', 'status', 'featured', 'catalog_visibility', 'purchasable', 'virtual', 'downloadable', 'downloads']


In [5]:
df= df.drop(['type', 'status', 'featured', 'catalog_visibility', 'purchasable', 'virtual', 'downloadable', 'downloads'], axis = 1)
df_anon = df_anon.drop(['version','Unnamed: 0.1', 'Unnamed: 0','parent_id', 'currency', 'prices_include_tax', 'discount_tax', 'shipping_tax', 'cart_tax', 'total_tax', 'tax_lines', 'fee_lines', 'currency_symbol', 'shipping.phone'], axis = 1)

In [6]:
df_anon['line_items']

0       [{'id': 4099, 'name': 'Infinity verde + 30 fil...
1       [{'id': 4096, 'name': 'Infinity grigio + 30 fi...
2       [{'id': 4093, 'name': 'Maglia Termica Slalom M...
3       [{'id': 4091, 'name': 'Infinity verde + 30 fil...
4       [{'id': 4089, 'name': 'Infinity mask rosso + 3...
                              ...                        
1407    [{'id': 19, 'name': 'Maglia Termica Training W...
1408    [{'id': 17, 'name': 'Maglia Termica Training B...
1409                                                   []
1410    [{'id': 14, 'name': 'Maglione Red Stripes', 'p...
1411    [{'id': 8, 'name': 'Calzamaglia Termica Silver...
Name: line_items, Length: 1412, dtype: object

In [7]:
df = df.drop(['description', 'short_description','images','price_html'], axis = 1) 
df

id                          name                             slug  \
0   3418                 Kit 30 filtri                    kit-30-filtri   
1   3401  Infinity azzurro + 30 filtri  infinity-mask-azzurro-30-filtri   
2   3387   Infinity grigio + 30 filtri   infinity-mask-grigio-30-filtri   
3   3351    Infinity verde + 30 filtri    infinity-mask-verde-30-filtri   
4   3397  Infinity arancio + 30 filtri  infinity-mask-arancio-30-filtri   
..   ...                           ...                              ...   
64  1369   T-shirt Fitness Camo Yellow      t-shirt-fitness-camo-yellow   
65  1362     T-shirt Fitness Camo Grey        t-shirt-fitness-camo-grey   
66  1355     T-shirt Fitness Camo Blue        t-shirt-fitness-camo-blue   
67  1348        3/4 Leggings Camo Grey   3-4-leggings-fitness-camo-grey   
68  1341      3/4 Leggings Camo Yellow         3-4-leggings-camo-yellow   

                                            permalink         date_created  \
0   https://archived.forceteksport.it/product/kit-...  2020-04-28T15:49:52   
1   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:36   
2   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:18   
3   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:09   
4   https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:08   
..                                                ...                  ...   
64  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:30   
65  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:30   
66  https://archived.forceteksport.it/product/t-sh...  2019-12-05T08:56:29   
67  https://archived.forceteksport.it/product/3-4-...  2019-12-05T08:56:28   
68  https://archived.forceteksport.it/product/3-4-...  2019-12-05T08:56:27   

       date_created_gmt        date_modified    date_modified_gmt      sku  \
0   2020-04-28T15:49:52  2021-08-10T14:37:42  2021-08-10T14:37:42  FS00056   
1   2020-04-28T10:00:36  2021-09-02T06:31:50  2021-09-02T06:31:50  FS00051   
2   2020-04-28T10:00:18  2021-09-24T08:24:08  2021-09-24T08:24:08  FS00053   
3   2020-04-28T10:00:09  2021-10-03T18:05:27  2021-10-03T18:05:27  FS00055   
4   2020-04-28T10:00:08  2021-09-04T05:55:29  2021-09-04T05:55:29  FS00052   
..                  ...                  ...                  ...      ...   
64  2019-12-05T08:56:30  2021-02-27T18:35:52  2021-02-27T18:35:52   D00009   
65  2019-12-05T08:56:30  2021-06-18T15:59:29  2021-06-18T15:59:29   D00010   
66  2019-12-05T08:56:29  2021-02-27T18:35:52  2021-02-27T18:35:52   D00008   
67  2019-12-05T08:56:28  2020-10-27T14:10:30  2020-10-27T14:10:30   D00017   
68  2019-12-05T08:56:27  2021-01-23T14:58:15  2021-01-23T14:58:15   D00016   

    price  ...                            variations  \
0   19.90  ...                          [3419, 3421]   
1   29.90  ...              [4802, 3404, 3403, 3402]   
2   29.90  ...                    [3390, 3389, 3388]   
3   29.90  ...                    [3354, 3353, 3352]   
4   29.90  ...              [4803, 3400, 3399, 3398]   
..    ...  ...                                   ...   
64  34.95  ...        [7868, 7869, 7870, 7871, 7872]   
65  34.95  ...        [8067, 8068, 8069, 8070, 8071]   
66  34.95  ...        [8072, 8073, 8074, 8075, 8076]   
67  34.95  ...  [7980, 7981, 7982, 7983, 7984, 7985]   
68  34.95  ...  [7986, 7987, 7988, 7989, 7990, 7991]   

                       related_ids  \
0   [1459, 1555, 1596, 1341, 1540]   
1   [1569, 1422, 1396, 1555, 1376]   
2   [1434, 1355, 1390, 1396, 1422]   
3   [1569, 1459, 1369, 1404, 1396]   
4   [1404, 1569, 1596, 1411, 1341]   
..                             ...   
64  [1589, 1390, 1562, 1355, 1376]   
65  [1589, 1376, 1417, 1533, 1422]   
66  [1369, 1427, 1589, 1376, 1383]   
67  [1548, 1404, 1596, 1555, 1434]   
68  [1569, 1562, 1362, 1459, 1540]   

                                            meta_data  \
0   [{'id': 34598, 'key': '_wpml_word_count'

'default_attributes' and attributes in the items df can also be used to extract info about sizes ordered and 'tags' can be used for gender identification

I would use line_items attribute in df_anon to match it with names and id's in the df.
 
 example: [{'id': 4099, 'name': 'Infinity verde + 30 filtri', 'product_id': 3351, 'variation_id': 3352, 'quantity': 1, 'tax_class': '', 'subtotal': '29.90', 'subtotal_tax': '0.00', 'total': '29.90', 'total_tax': '0.00', 'taxes': [], 'meta_data': [{'id': 39083, 'key': 'pa_size', 'value': 'l', 'display_key': 'Size', 'display_value': 'L'}, {'id': 39094, 'key': '_reduced_stock', 'value': '1', 'display_key': '_reduced_stock', 'display_value': '1'}], 'sku': 'FS00055.L', 'price': 29.9, 'parent_name': 'Infinity verde + 30 filtri'}] -> one can unwind the cell and extract matching attributes

In [8]:
df_anon['line_items'] = df_anon['line_items'].apply(eval)
df_anon['sku'] = df_anon['line_items'].apply(lambda x: x[0]['sku'] if x else None)
df_anon['product_name'] = df_anon['line_items'].apply(lambda x: x[0]['name'] if x else None)
df_anon['product_id'] = df_anon['line_items'].apply(lambda x: int(x[0]['product_id']) if x and x[0].get('product_id') else None)
df_anon['quantity'] = df_anon['line_items'].apply(lambda x: x[0]['quantity'] if x else None)

In [9]:
df_anon = df_anon.dropna(subset=['sku'])
df_anon

id     status      date_created     date_modified  discount_total  \
0     10367  completed   10/3/2021 18:05    10/4/2021 8:15            0.00   
1     10365  completed    9/24/2021 8:24    9/27/2021 7:44            0.00   
2     10363  completed   9/23/2021 11:24   9/23/2021 16:35            0.00   
3     10361  completed   9/14/2021 10:43    9/15/2021 7:38            0.00   
4     10359  completed     9/9/2021 9:47    9/10/2021 7:31            0.00   
...     ...        ...               ...               ...             ...   
1406   2943  completed   2/17/2020 22:34   2/20/2020 10:57            0.00   
1407   2806  completed   2/12/2020 16:33    2/14/2020 8:52            0.00   
1408   2686  completed   1/13/2020 13:14   1/13/2020 13:31           26.07   
1410   2656  completed  12/23/2019 13:41    3/3/2020 16:04           54.27   
1411   2651  completed  12/23/2019 12:17  12/23/2019 13:10           50.04   

      shipping_total   total  customer_id               order_key  \
0                8.9   38.80            0  wc_order_J5gjA5WJJF3rr   
1                0.0   59.80          364  wc_order_imifPBkjkzj7o   
2                0.0   83.35            0  wc_order_dvSnmV6DPbqC7   
3                8.9   38.80            0  wc_order_rV2J8azTl780O   
4                8.9   38.80            0  wc_order_NTwkbwTqjubFe   
...              ...     ...          ...                     ...   
1406             0.0  173.80           10  wc_order_wYu0MkTQKAgeN   
1407             0.0   86.90            0  wc_order_9KqecyJBHS2o4   
1408             0.0   60.83            9  wc_order_532nqvsija0up   
1410             0.0  126.63            8  wc_order_PsUU2rLQNrlfp   
1411             0.0  116.76            7  wc_order_23s9F6MExquZV   

     payment_method  ...  Fake_billing.phone Fake_shipping.first_name  \
0               cod  ...  WOQG33846662194699       SVNY16148140136361   
1            stripe  ...  TMUL91089863529617       RQVH21863526139131   
2            paypal  ...  BUTB52390212629269       CSQN67144071641919   
3            paypal  ...  YPBV68830205694380       TKFT05159506495260   
4            paypal  ...  OARY68449684287032       ETJP40556153503170   
...             ...  ...                 ...                      ...   
1406         stripe  ...  SJYR48849542615233       DKWM27350958177151   
1407         stripe  ...  RCOI13964143490186       DKWM27350958177151   
1408            NaN  ...  BCDE08611747311795       DKWM27350958177151   
1410         paypal  ...  BCDE08611747311795       DKWM27350958177151   
1411            NaN  ...  BCDE08611747311795       DKWM27350958177151   

     Fake_shipping.last_name Fake_shipping.company Fake_shipping.address_1  \
0         IPVW04117884001686    NGFU59411694115869      RQRT64695418368355   
1         AVGD15380084592939    NGFU59411694115869      WMPZ65551326895322   
2         NYTZ58980820015769    NGFU59411694115869      PDPW41517902757120   
3         FULX40410249147544    NGFU59411694115869      LSUA50241147302188   
4         MUDE47985565585748    NGFU59411694115869      ZYKC53909926722322   
...                      ...                   ...                     ...   
1406      YCIE22246655001549    NGFU59411694115869      OKZX24508433421792   
1407      YCIE22246655001549    NGFU59411694115869      OKZX24508433421792   
1408      YCIE22246655001549    NGFU59411694115869      OKZX24508433421792   
1410      YCIE22246655001549    NGFU59411694115869      OKZX24508433421792   
1411      YCIE22246655001549    NGFU59411694115869      OKZX24508433421792   

     Fake_shipping.address_2        sku                     product_name  \
0         REEY66120518219128  FS00055.L       Infinity verde + 30 filtri   
1         REEY66120518219128  FS00053.M      Infinity grigio + 30 filtri   
2         REEY66120518219128     D00006        Maglia Termica Slalom Man   
3         REEY66120518219128  FS00055.L       Infinity verde + 30 filtri   
4         REEY66120518219128  FS00050.L  Infinity

Joined the df's by product name extracted from the line column

In [10]:
df = df.rename(columns={'id':'product_id'})

In [11]:
all = pd.merge(df,df_anon, on = 'product_id', how = 'right')

In [12]:
all

product_id                             name  \
0           3351       Infinity verde + 30 filtri   
1           3387      Infinity grigio + 30 filtri   
2           1533        Maglia Termica Slalom Man   
3           3351       Infinity verde + 30 filtri   
4           3331  Infinity mask rosso + 30 filtri   
...          ...                              ...   
1406        1562         Maglia Termica Spear Man   
1407        1616    Maglia Termica Training White   
1408        1602    Maglia Termica Training Black   
1409        2224             Maglione Red Stripes   
1410        1589       Calzamaglia Termica Silver   

                                slug  \
0      infinity-mask-verde-30-filtri   
1     infinity-mask-grigio-30-filtri   
2          maglia-termica-slalom-man   
3      infinity-mask-verde-30-filtri   
4      infinity-mask-rosso-30-filtri   
...                              ...   
1406        maglia-termica-spear-man   
1407   maglia-termica-training-white   
1408   maglia-termica-training-black   
1409   maglione-red-stripes-forcetek   
1410      calzamaglia-termica-silver   

                                              permalink       date_created_x  \
0     https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:09   
1     https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:18   
2     https://archived.forceteksport.it/product/magl...  2019-12-05T08:56:48   
3     https://archived.forceteksport.it/product/infi...  2020-04-28T10:00:09   
4     https://archived.forceteksport.it/product/infi...  2020-01-29T10:00:41   
...                                                 ...                  ...   
1406  https://archived.forceteksport.it/product/magl...  2019-12-05T08:56:51   
1407  https://archived.forceteksport.it/product/magl...  2019-12-05T08:56:57   
1408  https://archived.forceteksport.it/product/magl...  2019-12-05T08:56:56   
1409  https://archived.forceteksport.it/product/magl...  2019-12-06T17:29:03   
1410  https://archived.forceteksport.it/product/calz...  2019-12-05T08:56:54   

       date_created_gmt_x      date_modified_x  date_modified_gmt_x    sku_x  \
0     2020-04-28T10:00:09  2021-10-03T18:05:27  2021-10-03T18:05:27  FS00055   
1     2020-04-28T10:00:18  2021-09-24T08:24:08  2021-09-24T08:24:08  FS00053   
2     2019-12-05T08:56:48  2021-09-23T11:26:05  2021-09-23T11:26:05   D00006   
3     2020-04-28T10:00:09  2021-10-03T18:05:27  2021-10-03T18:05:27  FS00055   
4     2020-01-29T10:00:41  2021-09-24T08:24:08  2021-09-24T08:24:08  FS00050   
...                   ...                  ...                  ...      ...   
1406  2019-12-05T08:56:51  2020-12-07T15:15:48  2020-12-07T15:15:48   D00001   
1407  2019-12-05T08:56:57  2021-06-01T10:14:25  2021-06-01T10:14:25   D00014   
1408  2019-12-05T08:56:56  2021-06-01T10:17:38  2021-06-01T10:17:38   D00011   
1409  2019-12-06T17:29:03  2021-01-13T16:51:49  2021-01-13T16:51:49   D00095   
1410  2019-12-05T08:56:54  2020-08-31T00:02:26  2020-08-31T00:02:26   D00022   

      price  ...  Fake_billing.email  Fake_billing.phone  \
0     29.90  ...  FBFX72905078821135  WOQG33846662194699   
1     29.90  ...  BNMS60884764650920  TMUL91089863529617   
2     43.45  ...  HEFQ73268729848592  BUTB52390212629269   
3     29.90  ...  OWQK75971927080614  YPBV68830205694380   
4     29.90  ...  YUFU37809794516827  OARY68449684287032   
...     ...  ...                 ...                 ...   
1406  43.45  ...  TNEN20111396652823  SJYR48849542615233   
1407  43.45  ...  ERVV21393195528266  RCOI13964143490186   
1408  43.45  ...  KEJH02132819969939  BCDE08611747311795   
1409  90.45  ...  NWCV38461262576518  BCDE08611747311795   
1410  39.90  ...  NVXL40949272696714  BCDE08611747311795   

      Fake_shipping.first_name  Fake_shipping.last_name Fake_shipping.company  \
0           SVNY16148140136361       IPVW04117884001686    NGFU59411694115869   
1           RQVH21863526139131       AVGD15380084592939    NGFU59411694115869   
2 

Now I want to get the item size if possible and remove sku duplicate column

In [13]:
all[['sku','size']] = all['sku_y'].str.split('.', expand=True)

In [14]:
all.drop(['permalink','slug','sku_x','sku_y','default_attributes','attributes','date_created_x','date_modified_x', 'product_name'], axis = 1)
#id_x = product_id

product_id                             name   date_created_gmt_x  \
0           3351       Infinity verde + 30 filtri  2020-04-28T10:00:09   
1           3387      Infinity grigio + 30 filtri  2020-04-28T10:00:18   
2           1533        Maglia Termica Slalom Man  2019-12-05T08:56:48   
3           3351       Infinity verde + 30 filtri  2020-04-28T10:00:09   
4           3331  Infinity mask rosso + 30 filtri  2020-01-29T10:00:41   
...          ...                              ...                  ...   
1406        1562         Maglia Termica Spear Man  2019-12-05T08:56:51   
1407        1616    Maglia Termica Training White  2019-12-05T08:56:57   
1408        1602    Maglia Termica Training Black  2019-12-05T08:56:56   
1409        2224             Maglione Red Stripes  2019-12-06T17:29:03   
1410        1589       Calzamaglia Termica Silver  2019-12-05T08:56:54   

      date_modified_gmt_x  price  on_sale  total_sales  download_limit  \
0     2021-10-03T18:05:27  29.90    False          291              -1   
1     2021-09-24T08:24:08  29.90    False          701              -1   
2     2021-09-23T11:26:05  43.45     True            3              -1   
3     2021-10-03T18:05:27  29.90    False          291              -1   
4     2021-09-24T08:24:08  29.90    False          392              -1   
...                   ...    ...      ...          ...             ...   
1406  2020-12-07T15:15:48  43.45     True           11              -1   
1407  2021-06-01T10:14:25  43.45     True            2              -1   
1408  2021-06-01T10:17:38  43.45     True            4              -1   
1409  2021-01-13T16:51:49  90.45     True            2               0   
1410  2020-08-31T00:02:26  39.90     True            1              -1   

      download_expiry                                         categories  ...  \
0                  -1  [{'id': 207, 'name': 'Infinity', 'slug': 'infi...  ...   
1                  -1  [{'id': 207, 'name': 'Infinity', 'slug': 'infi...  ...   
2                  -1  [{'id': 181, 'name': 'Drytek', 'slug': 'drytek...  ...   
3                  -1  [{'id': 207, 'name': 'Infinity', 'slug': 'infi...  ...   
4                  -1  [{'id': 207, 'name': 'Infinity', 'slug': 'infi...  ...   
...               ...                                                ...  ...   
1406               -1  [{'id': 181, 'name': 'Drytek', 'slug': 'drytek...  ...   
1407               -1  [{'id': 172, 'name': 'Donna', 'slug': 'donna'}...  ...   
1408               -1  [{'id': 181, 'name': 'Drytek', 'slug': 'drytek...  ...   
1409                0  [{'id': 186, 'name': 'Lifestyle', 'slug': 'lif...  ...   
1410               -1  [{'id': 181, 'name': 'Drytek', 'slug': 'drytek...  ...   

      Fake_billing.email  Fake_billing.phone Fake_shipping.first_name  \
0     FBFX72905078821135  WOQG33846662194699       SVNY16148140136361   
1     BNMS60884764650920  TMUL91089863529617       RQVH21863526139131   
2     HEFQ73268729848592  BUTB52390212629269       CSQN67144071641919   
3     OWQK75971927080614  YPBV68830205694380       TKFT05159506495260   
4     YUFU37809794516827  OARY68449684287032       ETJP40556153503170   
...                  ...                 ...                      ...   
1406  TNEN20111396652823  SJYR48849542615233       DKWM27350958177151   
1407  ERVV21393195528266  RCOI13964143490186       DKWM27350958177151   
1408  KEJH02132819969939  BCDE08611747311795       DKWM27350958177151   
1409  NWCV38461262576518  BCDE08611747311795       DKWM27350958177151   
1410  NVXL40949272696714  BCDE08611747311795       DKWM27350958177151   

     Fake_shipping.last_name Fake_shipping.company Fake_shipping.address_1  \
0         IPVW04117884001686    NGFU59411694115869      RQRT64695418368355   
1         AVGD15380084592939    NGFU59411694115869      WMPZ65551326895322   
2         NYTZ58980820015769    NGFU59411694115869      PDPW41517902757120   
3         FULX40410249147544    NGFU59411694115869      LSUA502411

The df 'all' gives us an extended list of items with sizes and with associated customer information. I can confirm that there is no info from the items df that could not be extracted from the attribute 'line_items'. Therefore, the merge is not crucial. It could be a bit less painful to just merge than to unwind the column, though

In [15]:
print(len(all['customer_id'].unique()))

375


Order id is unique to each row so it can be neglected for customerid purposes. CustomerID can be build from a combination of attributes like name, email, address etc. but it is not reliable since there one client can have multiple bank accounts or addresses. There is a customer_id column which has 375 observations but a significant number of them is 0, so it cannot be used.

To construct the customer_id, I filtered out by email:

In [17]:
all['cust_id'] = all.groupby(['Fake_billing.email'], sort = False).ngroup()
all['cust_id'].nunique()

1172

In [18]:
all['cust_id'].value_counts()

cust_id
77      8
289     5
98      4
502     4
83      4
       ..
440     1
441     1
442     1
443     1
1171    1
Name: count, Length: 1172, dtype: int64

## Data exploration

1. Most sold items (no masks or winter clothes)

In [19]:
mask = all['status'] == 'completed'
completed = all[mask]
exclusion_words = ['Termica', 'Infinity', 'Mask','Filtri']
filtered_data = completed[~completed['name'].str.contains('|'.join(exclusion_words), case=False, na=False, regex=True)]
top3 = filtered_data['name'].value_counts()
top3.head(3)

name
Leggings Geometry       4
T-shirt Orange Kick     3
Leggings Basic Black    3
Name: count, dtype: int64

2. Are leggings and camisoles sold jointly or individually the most?

In [20]:
# print(all['cart_hash'].nunique())
#there seem to be 609 instances of unique carts created

#get item ids
item = completed['slug'].str.contains('canotta')
unique_product_ids = completed.loc[item, 'product_id'].unique()
print(unique_product_ids)
#1383 2052 1645 2195 2193 1459

[1383 2052]


In [21]:
id_c = [1383, 2052]
id_l = [1645, 2195, 2193, 1459]


order_keys_c = all[all['product_id'].isin(id_c)]['order_key'].unique()
order_keys_l = all[all['product_id'].isin(id_l)]['order_key'].unique()
are_order_keys_same = set(order_keys_c) == set(order_keys_l)

In [22]:
print(are_order_keys_same)
# does not seem to be an overlap between canotte and leggings -> sold individually

False


3. Top channel of acquisition (Facebook Ads, Google Search, etc.) which generates the most sales, either in number or monetary value -> i didnt find this data

4. Best-selling color (excluding black and excluding face masks and winter/thermal)

In [23]:
#extract colors and male/female tag (need it later) into separate columns
import json
all['tags'] = all['tags'].apply(eval)
all['color'] = all['tags'].apply(lambda tags: next((tag['name'].lower() for tag in tags if 'color' in tag['name'].lower() or 'color' in tag['slug'].lower()), ''))
all['gender'] = all['tags'].apply(lambda tags: next((tag['name'].lower() for tag in tags if 'gender' in tag['name'].lower() or 'gender' in tag['slug'].lower()), ''))
all['gender'] = all['gender'].str.replace('gender_', '')

In [24]:
top = all['color'].value_counts()
top.head(5)

# yellow, blue, red

color
                1366
color_black       38
color_yellow       4
color_blue         2
color_red          1
Name: count, dtype: int64

5. Most profitable ad campaign (i.e. with the best ROI) -> i didnt find this data


6. Average cost of customer acquisition (Ads spend/Num. customer acquired) -> i didnt find this data

7. Are shop carts are composed mostly by 1, 2 or 3 products? -> average is 2 products/cart

In [25]:
cart_counts = all['cart_hash'].value_counts()
print(cart_counts.mean())

2.2922824302134646


8. Number of people who buy men’s and women’s products jointly (i.e. in the same order)?

In [26]:
mask = all['gender'].isin(['man', 'woman'])
grouped = all[mask].groupby('cust_id')['gender'].nunique()
filtered = grouped[grouped > 1].index.tolist()

In [27]:
len(filtered)


1

## Data grouping
1. Total revenue
2. Total profit -> impossible to calculate unless we know the company's expenses
3. Frequency of purchase (per customer?)
4. Number of different product IDs purchased
5. Acquisition channel (Facebook Ads, Google Search, etc.) -> no such info
6. Day from last purchase (count from the latest date available)
7. Days from the date of acquisition

1. Total revenue calculation

In [28]:
mask = all['status'] == 'completed'
total_sum = (all['total'] - all['shipping_total'])[mask].sum()
print(total_sum)

73221.2


3.Frequency of purchase (per customer?)

In [29]:
completed = all[mask]
cust_id_counts = completed['cust_id'].value_counts()
print(cust_id_counts)

cust_id
77      7
13      4
502     4
28      4
83      4
       ..
434     1
435     1
436     1
437     1
1171    1
Name: count, Length: 1125, dtype: int64


4. Number of different product IDs purchased

In [30]:
print(len(completed['product_id'].value_counts()))

39


6. Day from last purchase (count from the latest date available)

In [31]:
from datetime import datetime
all['date_paid_gmt'] = pd.to_datetime(all['date_paid_gmt'])
max_date_paid_gmt = all['date_paid_gmt'].max()
today = pd.to_datetime(datetime.today().date())
days_from_last_purchase = (today - max_date_paid_gmt).days

print(days_from_last_purchase)

839


In [32]:
order_number_max_date = all.loc[all['date_paid_gmt'].idxmax(), 'order_key']

print(order_number_max_date)

wc_order_J5gjA5WJJF3rr


7. Days from the date of acquisition - I am not sure I understood the question, but I supposed it was the last order's completion date

In [33]:
order_key_to_find = order_number_max_date
date_completed = pd.to_datetime(all.loc[all['order_key'] == order_key_to_find, 'date_completed_gmt'].values[0])
today = pd.to_datetime(datetime.today().date())
days_from_first_acuisition = (today - date_completed).days

print(days_from_first_acuisition)

839
